### PR TITLE
Let OrderedClientCollection track the summarizer client separately

### LINKED ISSUE #7279: Consider changing summarization election to avoid 2 summarizer stepping on each other (last Summary)
In https://github.com/microsoft/FluidFramework/pull/7247, I've tried to reduce number of cases where we have two summarizer clients running at the same time by ensuring that newly starting client always waits a bit (under normal circumstances) before actually starting to summarize, giving some amount of time for the outgoing client to produce last summary.

This does work, but it's probabilistic, and I do see in stress test that we hit summary nacks due to two summarizer clients running. This creates extra noise that humans need to look into, but also consumes resources from client and server for nothing.

Much better flow should be something like this:
1. Today, summarizer election only considers interactive (not summarizer container) clients. I propose we change that and always given preference to running summarizer client and elect it as summarizer. That means that first some parent (interactive) client is elected, but then election moves to its child (summarizing container).
    - Please see SummaryManager.delayBeforeCreatingSummarizer() logic for details and comments in there.
2. Parent recognizes that and does not cancel summarizer client, it keeps going as today.
3. When parent disconnects or closes, it would (as today) notify summarizing client that it needs to move to exit. But no new summarizer starts and that allows last Summary to work without any concerns.
4. (no change required) once last summary is over, summarizing client will move to exit and new summarizer would spawn.

If we go with this route, we can change commented out checks in RunWhileConnectedCoordinator.cancelled to be more assertive about this.onBehalfOfClientId checks.

### PR DESCRIPTION
**Pull request by pleath**
That allows us to postpone electing a new client while the existing summarizer is still doing work, e.g., doing a last summary.

### CODE REVIEW

File: packages/runtime/container-runtime/src/orderedClientElection.ts [hunk 1]
```diff
@@ -335,11 +342,18 @@ export class OrderedClientElection
...
             this._eligibleCount++;
-            if (this._electedClient === undefined) {
+            if (!this.hasSummarizer) {
                 // Automatically elect latest client
                 this.tryElectingClient(client, sequenceNumber);
             }
         }
+        else if (client.client.details.type === "summmarizer") {
+            // Track this as the summarizer client if there's already an elected client.
+            // If there's no elected client, this must be an on-demand summarizer.
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/orderedClientElection.ts:352]: I do not think this statement is true. I.e. consider this sequence:
- main client starts
- at some point it becomes latest client (maybe immediately on connection).
- It starts summarizer client.
- It gets nacked from connection (because it sends some op, and there are too many unsummarized ops"
- some time passes...
- summarizer container finally boots and connects. It proceeds to "last summary" process.

I think at the end of this sequence, there might be no elected client. So new client booting will ignore summarizer

Comment (vladsud) [reply packages/runtime/container-runtime/src/orderedClientElection.ts:352]: I think the more interesting question:
Does system resolve properly conflict when two clients show up, and each quickly get nacked (due to too many unsumarized ops), but each starts a summarizer container (as each of them was the oldest non-summarized client at some point)? I.e., once both summarizer container boots, the second one should realize first one is still there and exit, right? Or at least not to produce any summarize and wait for first one to exit. This would reduce number of errors in the system, including summarizers stepping on each other.

**Comment by vladsud** [comment]
I'd love to see UT cases added clearly covering this area, including conditions this PR addresses. I.e., have tests that fail before this PR and succeed with it, and clearly demonstrating benefits (and also reducing risks)

File: packages/runtime/container-runtime/src/orderedClientElection.ts (hunk 1 above)
Comment (pleath) [reply packages/runtime/container-runtime/src/orderedClientElection.ts:352]: I see your point. I think there's also a problem with a different kind of case, e.g., where an on-demand summarizer client has been created before we have an elected one. I was trying to solve the problem of not letting the on-demand summarizer be treated as the elected one by having _electedClient===undefined mean that we haven't done an election yet, but I don't think that works. I think things become simpler if we can detect that a given summarizer client was created on demand and not automatically. Maybe that could be exposed in ICapabilities.

Comment (pleath) [reply packages/runtime/container-runtime/src/orderedClientElection.ts:352]: I do have a question about the ordering of "addClient"/"removeClient" messages. If an interactive client spins up a (non-interactive) summarizer client and then gets nacked, is it true that the "addClient" message for the summarizer client will be processed before the "removeClient" message for the interactive client, to reflect the order in which things actually happened? Or is the ordering not clear or guaranteed?

**Comment by pleath** [comment]
> I'd love to see UT cases added clearly covering this area, including conditions this PR addresses. I.e., have tests that fail before this PR and succeed with it, and clearly demonstrating benefits (and also reducing risks)

Agreed. I wanted to have eyes on the fix itself first.

File: packages/runtime/container-runtime/src/orderedClientElection.ts (hunk 1 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/orderedClientElection.ts:352]: No, there is no guarantees here. Non-interactive client takes a while to load and connect, so interactive client may have already disconnected by that time. Today, we allow such summarizer to keep limping along if conditions are right (too many unsummarized ops), even though another summarizer (who is properly elected) could be showing up.

I think the best outcome here - all clients (interactive or not) are taken into account when calculating a summarizer. And if either parent (interactive) or child (non-interactive) client is elected, then corresponding child container keeps going with summaries until it exists (commits or fails to commit last summary).

**Comment by pleath** [comment]
I'll be adding tests, though I'm not sure how to test the original scenario in which a new summarizer is elected after the delay while the original summarizer is still doing last summary. Also, is it possible to remove the opsToBypassInitialDelay part of delayBeforeCreatingSummarizer with this change?

**Comment by pleath** [comment]
Converting to draft, since I've found a couple of problems with this version of the change.

**Comment by pleath** [comment]
FYI, this is ready for review. I can provide an explanation for the separate tracking of electedClient and electedParent if that would be helpful. (Thanks.)

**Comment by vladsud** [comment]
Any chance you can update description with a bit more info on what is changed, what is new design, etc.

**Comment by pleath** [comment]
> Any chance you can update description with a bit more info on what is changed, what is new design, etc.

Sure. I've added a long comment to the OrderedClientElection class definition that explains how the scheme in this change works. Basically, it tracks the summarizer client and parent client separately so it can handle the case where the parent client changes (e.g., too many ops without summary, so we elect a new parent) but a new summarizer shouldn't be spawned until the old summarizer client leaves the quorum.

File: packages/runtime/container-runtime/src/orderedClientElection.ts [hunk 2]
```diff
@@ -206,8 +208,10 @@ export interface IOrderedClientElectionEvents extends IEvent {
 export interface ISerializedElection {
     /** Sequence number at the time of the latest election. */
     readonly electionSequenceNumber: number;
-    /** Most recently elected client id. */
+    /** Most recently elected client id. May be interactive client or summarizer client. */
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/orderedClientElection.ts:211]: May be worth adding that this can be an interactive client only if its same as the parent client and its an indication that a summarizer client has to be spawned.

File: packages/runtime/container-runtime/src/orderedClientElection.ts [hunk 3]
```diff
@@ -216,6 +220,8 @@ export interface IOrderedClientElection extends IEventProvider<IOrderedClientEle
     readonly eligibleCount: number;
     /** Currently elected client. */
     readonly electedClient: ITrackedClient | undefined;
+    /** Currently elected parent, may be equal to electedClient */
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/orderedClientElection.ts:223]: Same comment as above. It may be worth adding when this happens and what happens next.

File: packages/runtime/container-runtime/src/orderedClientElection.ts (hunk 2 above)
Comment (pleath) [reply packages/runtime/container-runtime/src/orderedClientElection.ts:211]: Thanks. That's a good suggestion.

File: packages/runtime/container-runtime/src/orderedClientElection.ts (hunk 3 above)
Comment (pleath) [reply packages/runtime/container-runtime/src/orderedClientElection.ts:223]: Sounds good.

File: packages/runtime/container-runtime/src/orderedClientElection.ts [hunk 4]
```diff
@@ -262,12 +308,16 @@ export class OrderedClientElection
     ) {
         super();
         let initialClient: ILinkedClient | undefined;
+        let initialParent: ILinkedClient | undefined;
         for (const client of orderedClientCollection.getAllClients()) {
             this.addClient(client, 0);
             if (typeof initialState !== "number") {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/orderedClientElection.ts:314]: What does initialState = number imply? The comment above says "current sequence number at time of load if new" but I am not sure what it means.
If this client was just created / loaded how can it be the elected client or parent? Isn't that only possible if this is the very first "write" client in the document?

Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/orderedClientElection.ts:314]: Oh I get it. It's comparing the intialState's client and parent Id against all clients in the list and not against "this" client. My bad.

File: packages/runtime/container-runtime/src/orderedClientElection.ts [hunk 5]
```diff
@@ -288,28 +338,44 @@ export class OrderedClientElection
                 });
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/orderedClientElection.ts:338]: The "if" block basically checks whether the elected client as per the initial state in summary exists. If not, `this._electedClient` will be undefined. I have couple of questions here:
1. Does this trigger a re-election?
2. The initial state in summary may be out of date right? Basically, there may already be a new summarizer chosen.

File: packages/runtime/container-runtime/src/orderedClientElection.ts (hunk 4 above)
Comment (pleath) [reply packages/runtime/container-runtime/src/orderedClientElection.ts:314]: "initialState === "number"" means that we didn't read an initially elected client from the snapshot, so the "initialState" just represents a sequence number. It may be true that today this only happens when there was no snapshot to load from. But note that the initial elected client is not the new one. It's whatever client was elected at the time of the snapshot. The point here, I think, is that all clients should have their OrderedClientElection's in the same state. There may be a better way to accomplish that. The way we're using today doesn't seem foolproof.

File: packages/runtime/container-runtime/src/orderedClientElection.ts [hunk 6]
```diff
@@ -288,28 +338,44 @@ export class OrderedClientElection
...
-        this._electionSequenceNumber = sequenceNumber;
-        if (this._electedClient === client) {
-            return;
-        }
+        let change = false;
+        const isSummarizerClient = client?.client.details.type === summarizerClientType;
         const prevClient = this._electedClient;
-        this._electedClient = client;
-        this.emit("election", client, sequenceNumber, prevClient);
+        if (this._electedClient !== client) {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/orderedClientElection.ts:360]: From the big comment above, this function will be called in step iii right. These are the sequence of steps:
1. electedParent = A, electedClient = A'.
2. `tryElectingClient` is called with client = B which is not a summarizer client.
3. This function will update both electedParent and electedClient to B.
But that should not happen right? It should only update electedParent. electedClient should be updated when A' leaves the quorum and this is called again. Is this understanding correct?

File: packages/runtime/container-runtime/src/orderedClientElection.ts (hunk 5 above)
Comment (pleath) [reply packages/runtime/container-runtime/src/orderedClientElection.ts:338]: 1. I don't think a re-election will be triggered immediately, but if we process an op, and _electedClient is undefined, then we'll do a new election, so it will be triggered soon.
2. Yes, you're right. In the case where we went too many ops without summarizing and elected a new one, but the new one hasn't completed a summary yet at the time the new client is loaded, we could read the old (unresponsive) client's ID from the snapshot we use to load the new client. If the unresponsive client is still in the quorum, the new client will think the unresponsive client is the elected summarizer. I _think_ this is only a problem when the new client becomes the elected client but doesn't know it. I'm not sure that's possible.

File: packages/runtime/container-runtime/src/orderedClientElection.ts (hunk 6 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/orderedClientElection.ts:360]: This is a little hard to follow :-(. I like the big comment but I think we should add smaller comments from it in these functions as well so we can piece this info together.

Comment (pleath) [reply packages/runtime/container-runtime/src/orderedClientElection.ts:360]: tryElectingClient should only get called if _electedClient is undefined or if the current _electedClient is not a summarizer client. It shouldn't be called to replace a summarizer client unless the summarizer client is leaving the quorum (i.e., in OrderedClientElection.removeClient).

File: packages/runtime/container-runtime/src/orderedClientElection.ts (hunk 5 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/orderedClientElection.ts:338]: Thanks, that makes sense. This is really a corner case and should not happen often. I was worried if this can happen a lot and looks like that's not the case.

File: packages/runtime/container-runtime/src/orderedClientElection.ts [hunk 7]
```diff
@@ -318,9 +384,10 @@ export class OrderedClientElection
      * @param client - client to start checking
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/orderedClientElection.ts:384]: The function description needs updating.

File: packages/runtime/container-runtime/src/orderedClientElection.ts [hunk 8]
```diff
@@ -335,10 +402,16 @@ export class OrderedClientElection
     private addClient(client: ILinkedClient, sequenceNumber: number): void {
         if (this.isEligibleFn(client)) {
             this._eligibleCount++;
-            if (this._electedClient === undefined) {
-                // Automatically elect latest client
+            const isSummarizer = client.client.details.type === summarizerClientType;
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/orderedClientElection.ts:405]: nit: Consider calling this isNewClientSummarizer or isAddedClientSummarizer for readability.

File: packages/runtime/container-runtime/src/orderedClientElection.ts [hunk 9]
```diff
@@ -335,10 +402,16 @@ export class OrderedClientElection
...
+            const electedClientIsSummarizer = this._electedClient?.client.details.type === summarizerClientType;
+            // Note that we allow a summarizer client to supercede an interactive client as elected client.
+            if (this._electedClient === undefined || (isSummarizer && !electedClientIsSummarizer)) {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/orderedClientElection.ts:408]: nit: Making the second part of this if statement (!electedClientIsSummarizer && isSummarizer) will read better. Basically, the elected client is not summarizer but the added client is.

File: packages/runtime/container-runtime/src/orderedClientElection.ts [hunk 10]
```diff
@@ -335,10 +402,16 @@ export class OrderedClientElection
...
                 this.tryElectingClient(client, sequenceNumber);
             }
+            else if (this._electedParent === undefined && !isSummarizer) {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/orderedClientElection.ts:411]: Should this "else" be removed and this condition always be checked? Basically, is it possible that after the "if" condition is executed, _electedParent is undefined? If so, we should try electing the parent too right.

File: packages/runtime/container-runtime/src/orderedClientElection.ts [hunk 11]
```diff
@@ -352,9 +425,32 @@ export class OrderedClientElection
         if (this.isEligibleFn(client)) {
             this._eligibleCount--;
             if (this._electedClient === client) {
-                // Automatically shift to next oldest client
-                const nextClient = this.findFirstEligibleClient(this._electedClient.youngerClient);
-                this.tryElectingClient(nextClient, sequenceNumber);
+                // Removing the _electedClient. There are 2 possible cases:
+                if (this._electedParent !== client) {
+                    // 1. The _electedClient is a summarizer that we've been allowing to finish its work.
+                    // Let the _electedParent become the _electedClient so that it can start its own summarizer.
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/orderedClientElection.ts:431]: It might be helpful to add which step from the scenario in the big comment is this (steps v and vi). Similarly, other points in code correspond to the scenario.
However, this does have the disadvantage that these comments can become outdated if anything changes!

File: packages/runtime/container-runtime/src/orderedClientElection.ts [hunk 12]
```diff
@@ -352,9 +425,32 @@ export class OrderedClientElection
...
+                }
+                else {
+                    // 2. The _electedClient is an interactive client that has left the quorum.
+                    // Automatically shift to next oldest client.
+                    const nextClient = this.findFirstEligibleParent(this._electedParent?.youngerClient) ??
+                        this.findFirstEligibleParent(this.orderedClientCollection.oldestClient);
+                    this.tryElectingClient(nextClient, sequenceNumber);
+                }
+            }
+            else if (this._electedParent === client) {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/orderedClientElection.ts:445]: This condition is not needed. This can just be "else"

File: packages/runtime/container-runtime/src/orderedClientElection.ts [hunk 13]
```diff
@@ -352,9 +425,32 @@ export class OrderedClientElection
...
+                // Removing the _electedParent. Shift to the next oldest one, but do not replace the _electedClient,
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/orderedClientElection.ts:446]: This scenario is where the electedClient and electedParent are the same. Again, referring to the scenario in the comment, this should happen when the old summarizer has left and the electedParent is chosen as the electedClient as well. So, the next step would have been for the electedParent to spawn a summarizer client which becomes electedClient.
If this client is torn down, doesn't that mean we should select both electedParent and electedClient? I am confused by this comment here.

File: packages/runtime/container-runtime/src/orderedClientElection.ts [hunk 14]
```diff
@@ -364,23 +460,39 @@ export class OrderedClientElection
     }
 
     public incrementElectedClient(sequenceNumber: number): void {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/orderedClientElection.ts:462]: nit: Not part of this PR but this function should be called `electNextEligibleClient` or something that conveys the same.

File: packages/runtime/container-runtime/src/orderedClientElection.ts [hunk 15]
```diff
@@ -364,23 +460,39 @@ export class OrderedClientElection
...
-        const nextClient = this.findFirstEligibleClient(this._electedClient?.youngerClient);
-        this.tryElectingClient(nextClient, sequenceNumber);
+        const nextClient = this.findFirstEligibleParent(this._electedParent?.youngerClient) ??
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/orderedClientElection.ts:463]: `this.findFirstEligibleParent(this._electedParent?.youngerClient) ?? this.findFirstEligibleParent(this.orderedClientCollection.oldestClient)` is done a few times. What does the two calls do? Also, maybe this needs to be a function called something like `findNextEligibleParent`?

File: packages/runtime/container-runtime/src/orderedClientElection.ts [hunk 16]
```diff
@@ -364,23 +460,39 @@ export class OrderedClientElection
...
+            // The _electedClient is a summarizer and should not be replaced until it leaves the quorum.
+            // Changing the _electedParent will stop the summarizer.
+            this.tryElectingParent(nextClient, sequenceNumber);
+        }
     }
 
     public resetElectedClient(sequenceNumber: number): void {
-        const firstClient = this.findFirstEligibleClient(this.orderedClientCollection.oldestClient);
-        this.tryElectingClient(firstClient, sequenceNumber);
+        const firstClient = this.findFirstEligibleParent(this.orderedClientCollection.oldestClient);
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/orderedClientElection.ts:476]: This and the above function are almost similar except how they find the parent. What's the difference between the two?

File: packages/runtime/container-runtime/src/orderedClientElection.ts [hunk 17]
```diff
@@ -364,23 +460,39 @@ export class OrderedClientElection
...
+        else {
+            // The _electedClient is a summarizer and should not be replaced until it leaves the quorum.
+            // Changing the _electedParent will stop the summarizer.
+            this.tryElectingParent(firstClient, sequenceNumber);
+        }
     }
 
     public peekNextElectedClient(): ITrackedClient | undefined {
-        return this.findFirstEligibleClient(this._electedClient?.youngerClient);
+        return this.findFirstEligibleParent(this._electedParent?.youngerClient);
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/orderedClientElection.ts:488]: Should this also look at this.orderedClientCollection.oldestClient similar to how its done in other places when finding the next client?

File: packages/runtime/container-runtime/src/summarizerClientElection.ts [hunk 18]
```diff
@@ -144,5 +149,5 @@ export class SummarizerClientElection
     }
 
     public static readonly clientDetailsPermitElection = (details: IClientDetails): boolean =>
-        details.capabilities.interactive && details.type !== summarizerClientType;
+        details.capabilities.interactive || details.type === summarizerClientType;
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summarizerClientElection.ts:152]: Doesn't this essentially mean all clients?

File: packages/runtime/container-runtime/src/summaryManager.ts [hunk 19]
```diff
@@ -199,18 +205,23 @@ export class SummaryManager implements IDisposable {
                 return;
             }
 
+            // We transition to Running before requesting the summarizer, because after requesting we can't predict
+            // when the electedClient will be replaced with the new summarizer client.
+            // The alternative would be to let connectedState.clientId !== clientElection.electedClientId when
+            // state === Starting || state === Running.
+            assert(this.state === SummaryManagerState.Starting, 0x263 /* "Expected: starting" */);
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summaryManager.ts:212]: This assert moved to before requesting the summarizer. Do we expect to be in the starting state now? This seems a little risky so making sure.

File: packages/runtime/container-runtime/src/orderedClientElection.ts (hunk 10 above)
Comment (pleath) [reply packages/runtime/container-runtime/src/orderedClientElection.ts:411]: No, tryElectingClient will set _electedParent if the new client is interactive, so the "else" is useful.

File: packages/runtime/container-runtime/src/orderedClientElection.ts (hunk 12 above)
Comment (pleath) [reply packages/runtime/container-runtime/src/orderedClientElection.ts:445]: The condition does need to be checked. If the client that's being removed is neither _electedClient nor _electedParent, then we should do nothing here.

File: packages/runtime/container-runtime/src/orderedClientElection.ts (hunk 13 above)
Comment (pleath) [reply packages/runtime/container-runtime/src/orderedClientElection.ts:446]: This is where electedClient and electedParent are _not_ equal. If they were equal, we would have entered the "if" path. On the "else" path, the electedParent is being removed, so we're going to change electedParent _only_. That will cause SummaryManager on that client to call Summarizer.stop.

File: packages/runtime/container-runtime/src/orderedClientElection.ts (hunk 15 above)
Comment (pleath) [reply packages/runtime/container-runtime/src/orderedClientElection.ts:463]: The first call tries to find the oldest eligible client that is younger than _electedParent. There may not be one, so then it will start over with the oldest client in the quorum.

File: packages/runtime/container-runtime/src/orderedClientElection.ts (hunk 16 above)
Comment (pleath) [reply packages/runtime/container-runtime/src/orderedClientElection.ts:476]: incrementElectedClient starts its search at the next-oldest client after electedParent, on the assumption that all the older ones are bad. resetElectedClient ignores electedParent and goes directly to the oldest client in the quorum.

File: packages/runtime/container-runtime/src/summarizerClientElection.ts (hunk 18 above)
Comment (pleath) [reply packages/runtime/container-runtime/src/summarizerClientElection.ts:152]: I'm not sure. It might mean that today, though you could imagine other kinds of non-interactive clients besides summarizers.

File: packages/runtime/container-runtime/src/orderedClientElection.ts (hunk 17 above)
Comment (pleath) [reply packages/runtime/container-runtime/src/orderedClientElection.ts:488]: I don't think so. We seem to be looking for the client that would be elected by incrementElectedClient.

File: packages/runtime/container-runtime/src/summaryManager.ts (hunk 19 above)
Comment (pleath) [reply packages/runtime/container-runtime/src/summaryManager.ts:212]: Yes, we expect it to be in starting state until the state is changed to Running. The real effect here is making the change to Running state earlier. That's to guarantee that until we're in Running state, _electedParent === _electedClient. After the summarizer client is spawned, _electedClient will be set to the summarizer client when the addClient message is processed.

File: packages/runtime/container-runtime/src/summaryManager.ts [hunk 20]
```diff
@@ -137,9 +137,15 @@ export class SummaryManager implements IDisposable {
...
+        // enforce connectedState.clientId === clientElection.electedClientId. But once we're Running, we should
+        // only transition to Stopping when the electedParentId changes. Stopping the summarizer without
+        // changing the electedParent will just cause us to transition to Starting again.
         if (!this.connectedState.connected) {
             return { shouldSummarize: false, stopReason: "parentNotConnected" };
-        } else if (this.connectedState.clientId !== this.clientElection.electedClientId) {
+        } else if (this.connectedState.clientId !== this.clientElection.electedParentId ||
+            (this.state !== SummaryManagerState.Running &&
+                this.connectedState.clientId !== this.clientElection.electedClientId)) {
             return { shouldSummarize: false, stopReason: "parentShouldNotSummarize" };
```
[the remark this answers is not in the document]
Comment (pleath) [reply packages/runtime/container-runtime/src/summaryManager.ts:149]: I take it to mean that the parent should stop its summarizer, if it has one, because a different parent has been elected (or, with this change, because a different summarizer is still running).

File: packages/runtime/container-runtime/src/orderedClientElection.ts [hunk 21]
```diff
@@ -262,12 +308,16 @@ export class OrderedClientElection
...
         super();
         let initialClient: ILinkedClient | undefined;
+        let initialParent: ILinkedClient | undefined;
         for (const client of orderedClientCollection.getAllClients()) {
             this.addClient(client, 0);
             if (typeof initialState !== "number") {
                 if (client.clientId === initialState.electedClientId) {
                     initialClient = client;
                 }
+                if (client.clientId === initialState.electedParentId) {
```
[the remark this answers is not in the document]
Comment (pleath) [reply packages/runtime/container-runtime/src/orderedClientElection.ts:318]: Good point. I'll walk through it and make sure.

File: packages/runtime/container-runtime/src/summarizerClientElection.ts (hunk 18 above)
Comment (NicholasCouri) [reply packages/runtime/container-runtime/src/summarizerClientElection.ts:152]: Out of curiosity, as I'm trying to learn. Wouldn't it mean that you could have the same client being summarizer again, if he is the oldest one ?

File: packages/runtime/container-runtime/src/orderedClientElection.ts [hunk 22]
```diff
@@ -206,16 +208,26 @@ export interface IOrderedClientElectionEvents extends IEvent {
...
+    readonly electedParentId: string | undefined;
 }
 
 /** Contract for maintaining a deterministic client election based on eligibility. */
 export interface IOrderedClientElection extends IEventProvider<IOrderedClientElectionEvents> {
     /** Count of eligible clients in the collection. */
     readonly eligibleCount: number;
-    /** Currently elected client. */
+    /** Currently elected client id. This is either:
+     * 1. the interactive elected parent client, in which case electedClientId === electedParentId,
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/orderedClientElection.ts:225]: You should remove id from this and the next comment. elected client id -> elected client. electedClientId -> electedClient. Same for parent.

File: packages/runtime/container-runtime/src/orderedClientElection.ts (hunk 12 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/orderedClientElection.ts:445]: oh. I thought this else if for the inner if. My bad.

File: packages/runtime/container-runtime/src/orderedClientElection.ts (hunk 13 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/orderedClientElection.ts:446]: Again, I thought this is the else to the inner if :-(.

File: packages/runtime/container-runtime/src/summarizerClientElection.ts (hunk 18 above)
Comment (NicholasCouri) [reply packages/runtime/container-runtime/src/summarizerClientElection.ts:152]: Makes sense! Thanks!

File: packages/runtime/container-runtime/src/orderedClientElection.ts (hunk 16 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/orderedClientElection.ts:476]: I see. It looks like if there is no elected client, resetElectedClient is called to choose the oldest client as summarizer. But if for some reason, the elected client is misbehaving (like it has not summarized for quite some time), incrementElectedClient is called.
nit: It would be nice to add comments for these functions to explain these :-).

File: packages/runtime/container-runtime/src/orderedClientElection.ts (hunk 17 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/orderedClientElection.ts:488]: I see. This is only used to log when the current client is not summarizing.

File: packages/runtime/container-runtime/src/orderedClientElection.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (pleath) [reply packages/runtime/container-runtime/src/orderedClientElection.ts:318]: Good catch. I think this would leave electedParent undefined, which could cause us never to summarize. Added some code to catch this case.

**Review by agarwal-navin — COMMENTED** [review]
LGTM.

**Review by agarwal-navin — APPROVED** [review]
